src/lora_subnet.erl

```diff
@@ -0,0 +1,537 @@
+    {Lower, Upper} = netid_addr_range(NetID, NetIDList),
+    (Addr >= Lower) and (Addr < Upper).
+
+-spec var_net_class(netclass()) -> non_neg_integer().
+var_net_class(NetClass) when NetClass =< 7 ->
+    IDLen = id_len(NetClass),
+    case NetClass of
+        0 -> 0;
+        1 -> 2#10 bsl IDLen;
+        2 -> 2#110 bsl IDLen;
+        3 -> 2#1110 bsl IDLen;
+        4 -> 2#11110 bsl IDLen;
+        5 -> 2#111110 bsl IDLen;
+        6 -> 2#1111110 bsl IDLen;
+        7 -> 2#11111110 bsl IDLen
+    end.
+
+-spec var_netid(netclass(), netid()) -> non_neg_integer().
+var_netid(NetClass, NetID) when NetClass =< 7 ->
+    NetID bsl addr_len(NetClass).
+
+-spec the_netid(number() | binary()) -> netid().
+the_netid(DevNum) when erlang:is_number(DevNum) ->
+    the_netid(<<DevNum:32/integer-unsigned>>);
+the_netid(DevAddr) ->
+    Type = netid_type(DevAddr),
+    ID = parse_id(DevAddr, Type + 1, id_len(Type)),
+    NetID = ID bor (Type bsl 21),
+    NetID.
+
+-spec parse_netid(number() | binary()) -> {ok, netid()} | {error, invalid_netid_type}.
+parse_netid(DevNum) when erlang:is_number(DevNum) ->
+    parse_netid(<<DevNum:32/integer-unsigned>>);
+parse_netid(DevAddr) ->
+    try
+        Type = netid_type(DevAddr),
+        ID = parse_id(DevAddr, Type + 1, id_len(Type)),
+        {ok, ID bor (Type bsl 21)}
+    catch
+        throw:invalid_netid_type:_ ->
+            {error, invalid_netid_type}
+    end.
+
+-spec addr_bit_len(number() | binary()) -> 7 | 10 | 13 | 15 | 17 | 20 | 24 | 25.
+addr_bit_len(DevNum) when erlang:is_number(DevNum) ->
+    addr_bit_len(<<DevNum:32/integer-unsigned>>);
+addr_bit_len(DevAddr) ->
+    NetID = the_netid(DevAddr),
+    addr_len(netid_class(NetID)).
+
+-spec parse_id(binary(), non_neg_integer(), non_neg_integer()) -> netid().
+parse_id(DevAddr, PrefixLength, NwkIDBits) ->
+    <<DevAddrNum:32/integer-unsigned>> = DevAddr,
+    uint32(uint32(DevAddrNum bsl PrefixLength) bsr (32 - NwkIDBits)).
+
+-spec netid_type(number() | binary()) -> 0..7.
+netid_type(NetID) when erlang:is_number(NetID) ->
+    netid_type(<<NetID:32/integer-unsigned>>);
+netid_type(<<First:8/integer-unsigned, _/binary>>) ->
+    netid_type(First, 7).
+
+-spec netid_type(non_neg_integer(), non_neg_integer()) -> 0..7.
+netid_type(_, -1) ->
+    throw(invalid_netid_type);
+netid_type(Prefix, Index) ->
+    case Prefix band (1 bsl Index) of
+        0 -> 7 - Index;
+        _ -> netid_type(Prefix, Index - 1)
+    end.
+
+-spec nwk_addr(devaddr()) -> nwkaddr().
+nwk_addr(DevAddr) ->
+    NetID = the_netid(DevAddr),
+    AddrBitLen = addr_len(netid_class(NetID)),
+    IgnoreLen = 32 - AddrBitLen,
+    DevAddr2 = <<DevAddr:32/integer-unsigned>>,
+    <<_:IgnoreLen, NwkAddr:AddrBitLen/integer-unsigned>> = DevAddr2,
+    NwkAddr.
+
+-spec netid_addr_range(netid(), [netid()]) -> {non_neg_integer(), non_neg_integer()}.
+netid_addr_range(NetID, NetIDList0) ->
+    FoundNetID = lists:any(fun(X) -> X == NetID end, NetIDList0),
+    case FoundNetID of
```

Comment:
Yes, false will be the most common case as we'll likely process packets that belong to non-Helium networks.  What is your suggestion for this case?

Reply:
ok.  That's fine. 

(Otherwise, there's computational overlap with `any()` followed by `takewhile()`.  Since that's the less traveled arm, nothing to do here unless we *measure* there being a performance bottleneck.  That's all I was thinking.)

Reply:
Yes, takewhile is the (much less) less traveled arm.  Also the netid list will be quite short, on the order of 1 to 5 items.